Add input for custom image url

bigggmoustache opened the pull request.

DoomScrollingTheYellowPages committed:
Add input for custom image url

bigggmoustache commented:
All of this is very new to me so if I did something you've requested wrong I apologize and look forward to correction.

jcreek commented:
Thank you for contributing to this project! I'll leave more specific feedback on the code itself, but in terms of the Pull Request, I'd recommend using `conventional commits` for your commit messages and PR titles. You can read more about this [here](https://www.jcreek.co.uk/dev-team/git/?h=conven#git-messages) but for this PR I would personally have had this commit message: 

`feat(16): Add input for custom image url`

This tells other devs that: 
- It's a feature, not a bug fix, chore or other change 
- It's related to Github Issue number 16, and on Github itself it directly links the PR to the issue, enabling it to be automatically closed when the PR is completed
- It clearly and concisely explains what has been changed 

Your original commit message `Add input for custom image url` meets this last point perfectly, so my only feedback is really the `feat(16): ` part - nice work! 🤜

jcreek commented on src/index.ts:
```diff
@@ -67,6 +67,14 @@ let data = [];
+  imageUrlForSpinner = imageUrlInput.value;
+}
+
+const button2 = document.getElementById('imageUrlSetter');
```
This variable name could be more verbose (clearer) - for example, it could be called `imageUrlSetterButton`, which would enable other devs to immediately know what it is.

bigggmoustache replied:
Sorry for this, I was actually just being cheeky because you left the 'startSpinning' button named 'button' lol. Would you like me to rename it as well?

jcreek replied:
Really good spot! Please do rename that one 😂

jcreek commented on src/styles/main.scss:
```diff
@@ -62,6 +62,14 @@ body {
   align-content: center;
   gap: 25px;
 }
+#imageUrl-container {
```
The input is very small, not making use of all available space. For example, in a desktop web browser, the URL field is very small, and there is a lot of redundant empty space on either side of the container. 

<img width="915" alt="image" src="https://github.com/jcreek/SelectionWheel/assets/8179928/fba9b660-38cc-425d-96d9-37e02eaabaf3">

On a mobile device it would be great to see the text input on its own line, with the button appearing below it, similar to what happens with the main textarea. 

<img width="330" alt="image" src="https://github.com/jcreek/SelectionWheel/assets/8179928/136f59a0-dcc2-4f8f-993c-3fbc09e22252">

jcreek replied:
This looks great now - nice work!

jcreek commented on src/template.html:
```diff
@@ -49,7 +49,10 @@
-
+        <div id="imageUrl-container">
+            <input type="text" id="imageUrlInput" name="imageUrlText">
+            <button id="imageUrlSetter" class ="button button-green" role="button">Set imageUrl</button>
```
There's an extra space here in `class ="`, it should be `class="`

jcreek commented on src/template.html:
```diff
@@ -49,7 +49,10 @@
             <button id="startSpinning" class="input-items button button-green" role="button">Start spinning</button>
         </div>
-
+        <div id="imageUrl-container">
```
This markup is great! It's simple and clear, with good well-named `id`s 🙂

jcreek commented on src/index.ts:
```diff
@@ -308,11 +318,15 @@ function startSpinning() {
 const button = document.getElementById('startSpinning');
 button.onclick = startSpinning;
 
+
```
These two blank lines are unnecessary and should be removed

jcreek commented on src/index.ts:
```diff
@@ -67,6 +67,14 @@ let data = [];
+const imageUrlInput = document.getElementById('imageUrlInput') as HTMLInputElement;
+
+function imageUrlSetter() {
+  imageUrlForSpinner = imageUrlInput.value;
```
In its current form, users are able to copy and paste a full URL, including the querystring, to share a customised selection wheel. It would be great for this function to also add the querystring for the image URL that the user has selected. 

Elsewhere in this file, we have this code: 

```ts
const urlParams = new URLSearchParams(window.location.search);
if (urlParams.has('imgUrl')) {
  imageUrlForSpinner = urlParams.get('imgUrl');
}
```

This enables us to get the imgUrl parameter from the querystring. 

Here, you're getting the url from the input field, so you could write some similar code within this function to set the querystring imgUrl parameter.

jcreek replied:
It would also be nice to populate the `imageUrlInput` text input field with the url from the querystring, if there is one. This could be done when we run `imageUrlForSpinner = urlParams.get('imgUrl');`

bigggmoustache replied:
I wasn't exactly sure how to go about this so I had the first function load a new window.location.href, and the second function is left as originally written. A notification that the page would reload seems appropriate, would you like me to add that below the input field? Everything requested is done. 😄 

Also sorry if the commit > tag stuff still isn't right (when I do it), I'm using fork and though I know how I'm not sure what'll come out of it.

jcreek replied:
Don't worry about changing the `window.location` - just updating the querystring is sufficient as people can copy it from their URL bar in the browser, if they want to share it, so no notification about reloading is necessary 🙂

No worries - just have a go!

bigggmoustache replied:
Howdy, I made the requested changes. Thank you again for putting something like this out there.  I've absolutely loathed git since I started learning to program and my interaction with you is probably the most useful engagement with it I've had.

I look forward to submitting more pr's!

jcreek replied:
Almost there now! This version of the code will do two undesirable things:

1. Reload the page
2. Add an image url to the end of any existing image urls

The second thing is particularly bad, as it introduces a bug. A user can set a URL, then change it, and then two URLs will show in the querystring. 
<img width="1029" alt="image" src="https://github.com/jcreek/SelectionWheel/assets/8179928/9b2e05c1-598b-442f-9423-8f9b00b0edb1">

Instead, I'd recommend using something like the below. I haven't checked my syntax so you'll need to make sure it's right, but this should either modify the imgUrl or add it if there isn't one, without any duplicates and without reloading the page 🙂

```ts
const urlParams = new URLSearchParams(window.location.search);
urlParams.set(imageUrlForSpinner);
```

jcreek reviewed (changes requested):
Over all this is a really good PR! You have successfully added the functionality as requested in the original issue 🤜

I have left a few comments for enhancements and minor improvements such as changing variable names or removing whitespace. If there's anything that's not clear, please feel free to reply with any questions you have!

When adding further commits to this PR, please try to use the conventional commits I mentioned previously 🙂

jcreek commented:
Once you've made all your changes, you'll need to do a git push and they will automatically be added to this PR.

DoomScrollingTheYellowPages committed:
feat(16): Add input for custom image url

When applied this commit will:

-template.html
--Line 54: Remove extra space after `class`.

-index.ts
--Line 74: Set window url to contain query paramater via.
--Line 76-84: Rename `button2` to `imageUrlSetterButton` and moved to line 84. Set the current URL to include query params, set the existing input field to show the chosen value, and use the code moved here from lines 339-343.
--Line 328-329: Remove empty lines.

-styles/mainscss
--Add parameters to `#imageUrl-container` and `#imageUrlInput` styles that scale up with page size, wrap components when page too small, and set the input field to match the size of its corresponding button.

jcreek commented on src/index.ts:
```diff
@@ -325,8 +332,6 @@ function startSpinning() {
 const button = document.getElementById('startSpinning');
```
It'd be great if you could rename this one to something more suitable

bigggmoustache replied:
Done and done! I just realized I don't need to keep the commit names matching the issue name and should actually describe my commit lol. Sorry about that!

jcreek replied:
Github is not showing any changes to this button variable - have you pushed your latest changes?

jcreek replied:
Github is not showing any changes to this button variable - have you pushed your latest changes?

bigggmoustache replied:
I forgot to push after commit, it's there now.

DoomScrollingTheYellowPages committed:
feat(16): Add input for custom image url

In this commit:

'button' set to 'startSpinningButton'
